solutions/USACO open16-bull-in-china-shop.mdx

```diff
@@ -0,0 +1,135 @@
+
+Once we have shifted the pieces, if there are two '#' in the same place, or if the result of superimposing the pieces is not the same as the original figurine ($\texttt{pieces[0]}$), then this shift will not work.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(K^2 \cdot N^6)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <cstdio>
+#include <vector>
+using namespace std;
+bool check(int piece, int x, int y);
+
+int n;
+bool pieces[11][8][8];
+vector<int> side[11];
+int main() 
+{
+	freopen("bcs.in", "r", stdin);
+	freopen("bcs.out", "w", stdout);
+	int k;
+	cin >> n >> k;
+	char c;
+	int left, right, top, bottom;
+	for (int i = 0; i <= k; i++)
+	{
+		left = n - 1;
+		right = 0;
+		top = n - 1;
+		bottom = 0;
+		for (int j = 0; j < n; j++)
+		{
+			for (int l = 0; l < n; l++)
+			{
+				cin >> c;
+				pieces[i][j][l] = (c == '#');
+				//find the sides of the piece
+				if (pieces[i][j][l])
+				{
+					bottom = max(bottom, j);
+					top = min(top, j);
+					right = max(right, l);
+					left = min(left, l);
+				}
+			}
+		}
+		side[i] = {left, right, top, bottom};
+	}
+	
+	//try all the pieces and shifts to find the correct one
+	for (int i = 1; i <= k; i++)
+	{
+		for (int j = i + 1; j <= k; j++)
+		{
+			for (int idx = side[i][3] - n; idx <= side[i][2]; idx++)
+			{
+				for (int idy = side[i][1] - n; idy <= side[i][0]; idy++)
+				{
+					for (int jdx = side[j][3] - n; jdx <= side[j][2]; jdx++)
+					{
+						for (int jdy = side[j][1] - n; jdy <= side[j][0]; jdy++)
+						{
+							bool good = true;
+							for (int x = 0; x < n; x++)
+							{
+								for (int y = 0; y < n; y++)
+								{
+									bool ipiece = check(i, x + idx, y + idy);
+									bool jpiece = check(j, x + jdx, y + jdy);
+									//two '#' are in the same place
+									if (ipiece && jpiece)
+									{
+										good = false;
+										break;
+									}
+									//the result doesn't match the figurine
+									if (pieces[0][x][y] != (ipiece || jpiece))
+									{
+										good = false;
+										break;
+									}
+								}
+								if (!good)
+								{
+									break;
+								}
+							}
+							if (good)
+							{
+								cout << i << " " << j << endl;
+								return 0;
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+//check if a piece is in bounds and is '#'
+bool check(int piece, int x, int y)
+{
+	return x >= 0 && x < n && y >= 0 && y < n && pieces[piece][x][y];
+}
```

Comment:
The code looks good! Although I prefer don't like to put "{" in a new line :P

Reply:
Nevertheless wait for more reviews, because as I said before maybe we are missing something about the statement or official solution.